include/mpp/algorithm/back_substitution.hpp

```diff
@@ -0,0 +1,175 @@
+/**
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+
+ *   http://www.apache.org/licenses/LICENSE-2.0
+
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+#pragma once
+
+#include <mpp/detail/types/algo_types.hpp>
+#include <mpp/detail/utility/algorithm_helpers.hpp>
+#include <mpp/detail/utility/cpo_base.hpp>
+#include <mpp/utility/square.hpp>
+#include <mpp/matrix.hpp>
+
+#include <utility>
+
+namespace mpp
+{
+	namespace detail
+	{
+		template<typename To, std::size_t RowsExtent, std::size_t ColumnsExtent>
+		inline auto back_subst_on_buffer_unchecked(auto a, auto b, std::size_t n)
+		{
+			using x_matrix_t = matrix<default_floating_type, RowsExtent, ColumnsExtent>;
+			using x_buffer_t = typename x_matrix_t::buffer_type;
+			auto x_buffer    = x_buffer_t{};
+
+			// @TODO: Any way to make this utilize push_back?
+			allocate_buffer_if_vector(x_buffer, n, 1, default_floating_type{});
+
+			/**
+			 * Implementation of back substitution from
+			 * https://www.gaussianwaves.com/2013/05/solving-a-triangular-matrix-using-back-backward-substitution/
+			 */
+			for (auto row = n; row > std::size_t{}; --row)
+			{
+				const auto row_index = row - 1;
+
+				auto result = static_cast<default_floating_type>(b[index_2d_to_1d(1, row_index, 0)]);
+
+				for (auto column = n - 1; column > row_index; --column)
+				{
+					result -= a[index_2d_to_1d(n, row_index, column)] * x_buffer[column];
+				}
+
+				result /= a[index_2d_to_1d(n, row_index, row_index)];
```

Comment:
Same thing with forward substitution: Could there be cases where we're dividing by 0 here?